### PR TITLE
[MPS] Ternary elementwise specializations: inner_contiguous, inner_strided, dense ILP, 2D strided

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
# [MPS] Ternary elementwise specializations: inner_contiguous, inner_strided, dense ILP, 2D strided

Part of #187455 (migrating remaining MPSGraph elementwise ops to Metal kernels). Shared-infra PR: brings the ternary exec path up to parity with the binary path's specializations and adds one new lineage member, so the pending consumer PRs (`where`, `prelu`, the loss-op tails in #187557) get their noncontiguous fast paths from one place instead of per-op kernels.

## What's added

**Kernels (`c10/metal/indexing.h`):**
- `ternary_dense_ilp` — mirrors `binary_dense_ilp`; strictly opt-in per call site via the new `ilp_threshold` parameter.
- `ternary_inner_contiguous` — mirrors `binary_inner_contiguous` (every operand unit-stride in dim 0; ILP tile over vectorizable typed loads).
- `ternary_inner_strided` + `_cast` — **new**: decodes the outer coordinate once per thread, then walks `ILP_PER_THREAD` elements of the innermost run by each operand's constant dim-0 byte stride. Stride-0 inputs (broadcast bounds/conditions, 0-dim operands) just re-read one address, so this single kernel covers the broadcast, 0-dim, and mixed-dtype layouts without per-pattern kernels. There is deliberately no `inner_contiguous_cast`: runtime-typed loads can't vectorize, so `inner_strided_cast` is the one cast-side fast path.
- `ternary_strided`/`ternary_strided_cast` move from 1D to 2D dispatch (`thread_pos.x` = innermost coordinate), saving one div+mod per element — same as the unary precedent.

**Dispatch (`exec_ternary_kernel`):** merged signature `(iter, name, alpha, scalar_arg_type, ilp_threshold)` folds in the host plumbing carried by #189420 (alpha) and #187557 (ILP); both rebase to consumer-only diffs after this lands. Specialized flavors all require `!alpha`. `PYTORCH_TERNARY_FORCE_FLAVOR` (`scalar|strided|ilp|inner_contiguous|inner_strided`) pins a flavor for A/B benching; a force bypasses profitability gates, never correctness gates.

**Stacked on #189624** (two `exec_ternary_kernel` dispatch bugfixes, split out per the bugfix-first convention) — review only the top commit here; the base commit is that PR.

**Instantiation growth:** the blanket ternary macro goes from 5 to 8 host_names per (op, dtype); clamp (9 dtypes, the only registered ternary op) gains +36 kernel instantiations, measured +452 KB on `libtorch_cpu.dylib` including the new host dispatch code.

## Gate design (the interesting decision)

`inner_strided` is *correct* for any strided layout but only *profitable* in two families, and the gate selects exactly those: (a) unit-or-broadcast inner layouts (`isInnerUnitOrBroadcast`: every operand's inner stride is 0 or its element size), where the loads have locality; and (b) **every cast layout** — cast loads pay a per-element runtime-type switch that dominates the gather cost, so the amortized outer decode wins even without locality (measured 1.07× on clamp half-vs-float with a transposed operand, 1.48× on `where` with a transposed operand). Non-cast layouts outside (a) — e.g. a transposed operand — measure 4–7% *behind* the 2D strided kernel and stay generic; that negative measurement is in the table on purpose. `INNER_STRIDED_MIN_EXTENT = 8` from the sweep below.

## Perf

`benchmarks/mps/bench_ternary_infra.py` (in-tree), `torch.utils.benchmark.Timer`, K-iteration amortized loop behind one sync, 3 interleaved trials, GPU-utilization-guarded, M-series. clamp with tensor bounds (the only trunk ternary consumer).

**Forced A/B within this build** (B = new flavor, A = 2D strided; speedup = A/B):

| case | dtype | A ms | B ms | speedup |
|---|---|---:|---:|---:|
| channel broadcast [64,64,56,56] × [1,64,1,1] → inner_strided | f32 | 23.05 | 8.06 | **2.86×** |
| channel broadcast | f16 | 22.96 | 6.91 | **3.32×** |
| unit-inner broadcast [4096,4096] × [4096] → inner_contiguous | f32 | 20.02 | 10.32 | **1.94×** |
| unit-inner broadcast | f16 | 19.98 | 5.26 | **3.80×** |
| 0-dim tensor bounds [2^22] → inner_strided | f32 | 5.43 | 4.93 | 1.10× |
| 0-dim tensor bounds | f16 | 5.37 | 2.21 | **2.43×** |
| transposed x.t() (no broadcast) → inner_strided **(gated out)** | f32 | 30.07 | 31.30 | 0.96× |
| transposed | f16 | 26.08 | 27.97 | 0.93× |
| transposed **cast** (half x.t() vs float bounds) → inner_strided **(selected: cast rule)** | f16/f32 | 7.78 | 7.24 | **1.07×** |

**Sweep** (x[outer,inner], bounds[outer,1], numel 2^22, f32): inner_strided wins 1.89–2.21× at every extent from 8 to 4096 → `INNER_STRIDED_MIN_EXTENT = 8`.

**Unforced dispatch vs trunk control build** (same machine, alternating runs ×3; control = trunk checkout @ 3b923b1, which already dispatched clamp through the ternary Metal path — the ternary kernels are unchanged between 3b923b1 and this PR's base):

| case | trunk ms | this PR ms | speedup |
|---|---:|---:|---:|
| channel broadcast f32 | 25.65 | 7.94 | **3.23×** |
| channel broadcast f16 | 25.70 | 7.10 | **3.62×** |
| unit-inner broadcast f32 | 23.11 | 10.24 | **2.26×** |
| scalar-bounds f32 | 3.47 | 2.45 | **1.42×** |
| transposed f32 | 30.71 | 30.38 | 1.01× (no regression) |
| dense f32 | 4.90 | 4.87 | 1.01× (no change) |

Routing proof: unforced channel-broadcast matches forced `inner_strided` (8.03 vs 8.07 ms), not forced `strided` (23.3 ms). Cast-transposed routing proof: unforced 7.236 ms == forced inner_strided 7.239 ms, vs forced strided 7.776 ms.

## Tests

- `test_clamp_tensor_bounds_out_dtype` — regression test for bug fix 2 (fails on trunk at pipeline creation).
- `test_clamp_tensor_bounds_flavor_parity` — forces every flavor across the layouts it serves: ragged tails (numel % 4), inner extents 15/16/17 around the tile boundary, transposed, 0-dim bounds, mixed dtype; CPU-parity per flavor.
- Full MPS suite (`run_test.py --mps --keep-going`): sole failure is `TestMetalRewritePass::test_conv` (mobile prepack rewrite), which fails identically on this machine at the base revision and on every branch of this campaign.

## Diff size

**This PR only** — `3e6a3ccd4e6` (#189624 head) → `4e7ed4befd1`:

5 files changed, **+952/−39**.

**Full stack (cumulative, as shown by GitHub against `main`)**:

5 files changed, **+975/−40**. GitHub's Files changed view includes #189624; reviewers of
this PR should review the parent-to-head range above.

### This PR only by file

| file | Δ |
|---|---|
| `c10/metal/indexing.h` | +309/−13 (new kernels + registrations, strided 1D→2D) |
| `aten/src/ATen/native/mps/OperationUtils.mm` | +153/−25 (merged exec, flavor selection, gates) |
| `aten/src/ATen/native/mps/MetalShaderLibrary.h` | +10/−1 (merged signature) |
| `test/test_mps.py` | +35/−0 (flavor-parity coverage) |
| `benchmarks/mps/bench_ternary_infra.py` | +445/−0 (new, in-tree bench) |
| **total** | **+952/−39** |

## Follow-ups (deliberately out of scope)

- Binary `inner_strided` generalization (prelu's NCHW weight layouts) — severed into its own PR with a metallib-size and bench proof.
- `ternary_alpha_dense_ilp` — owned by #189420 (addcmul/addcdiv) once this lands.
- Consumer opt-ins: `where` (parked branch becomes a thin consumer), BCE noncontig/mixed-dtype tails in #187557, mse ILP opt-in in #187553.
- Binary strided 1D→2D, mid-size fused-reduce tuning (tracked separately).

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -6603,6 +6603,57 @@ def helper(input_shape, bound_shape):
...
+            expect = torch.clamp(x, min=min_t, max=max_t)
+            for flavor in ["scalar", "strided", "ilp", "inner_contiguous", "inner_strided"]:
+                os.environ["PYTORCH_TERNARY_FORCE_FLAVOR"] = flavor
+                try:
+                    got = torch.clamp(x.to("mps"), min=min_t.to("mps"), max=max_t.to("mps"))
+                finally:
+                    del os.environ["PYTORCH_TERNARY_FORCE_FLAVOR"]
+                self.assertEqual(got.cpu(), expect, msg=f"flavor={flavor}")
+
+        for dtype in [torch.float32, torch.float16]:
```
Comment (Skylion007) [inline test/test_mps.py:7124]: self.subTest the dtypes here

Comment (anagnorisis2peripeteia) [reply test/test_mps.py:7124]: Done — each case now runs under `self.subTest(dtype=..., flavor=..., shapes=...)` so a failure names the exact combination and the remaining cases still run. (Same push also widens the inner_strided selection gate to all cast layouts with measurements in the updated description.)

**Comment by anagnorisis2peripeteia** [comment]
@Skylion007 any chance you can take another look please?
